plugins/drivers/utils/eventer.go

```diff
@@ -46,56 +51,86 @@ func NewEventer(stop chan struct{}) *Eventer {
 func (e *Eventer) eventLoop() {
 	for {
 		select {
-		case <-e.stop:
-			for _, stream := range e.streamers {
-				close(stream.ch)
-			}
+		case <-e.ctx.Done():
+			close(e.events)
 			return
 		case event := <-e.events:
-			e.RLock()
-			for _, stream := range e.streamers {
-				stream.send(event)
+			e.consumersLock.RLock()
+			for _, consumer := range e.consumers {
+				consumer.send(event)
 			}
-			e.RUnlock()
+			e.consumersLock.RUnlock()
 		}
 	}
 }
 
-type eventStreamer struct {
+type eventConsumer struct {
 	timeout time.Duration
 	ctx     context.Context
 	ch      chan *drivers.TaskEvent
+	logger  hclog.Logger
 }
 
-func (s *eventStreamer) send(event *drivers.TaskEvent) {
+func (c *eventConsumer) send(event *drivers.TaskEvent) {
 	select {
-	case <-time.After(s.timeout):
-	case <-s.ctx.Done():
-	case s.ch <- event:
+	case <-time.After(c.timeout):
+		c.logger.Warn("timeout sending event", "task_id", event.TaskID, "message", event.Message)
+	case <-c.ctx.Done():
+	case c.ch <- event:
 	}
 }
 
-func (e *Eventer) newStream(ctx context.Context) <-chan *drivers.TaskEvent {
-	e.Lock()
-	defer e.Unlock()
+func (e *Eventer) newConsumer(ctx context.Context) *eventConsumer {
+	e.consumersLock.Lock()
+	defer e.consumersLock.Unlock()
 
-	stream := &eventStreamer{
+	consumer := &eventConsumer{
 		ch:      make(chan *drivers.TaskEvent),
 		ctx:     ctx,
 		timeout: DefaultSendEventTimeout,
+		logger:  e.logger,
 	}
-	e.streamers = append(e.streamers, stream)
+	e.consumers = append(e.consumers, consumer)
 
-	return stream.ch
+	return consumer
 }
 
 // TaskEvents is an implementation of the DriverPlugin.TaskEvents function
 func (e *Eventer) TaskEvents(ctx context.Context) (<-chan *drivers.TaskEvent, error) {
-	stream := e.newStream(ctx)
-	return stream, nil
+	consumer := e.newConsumer(ctx)
+	go e.handleConsumer(consumer)
+	return consumer.ch, nil
+}
+
+func (e *Eventer) handleConsumer(consumer *eventConsumer) {
```

Comment:
Lets handle cleanup in the event loop so we only have one go routine? As you loop through the consumers, on emit event we can cleanup and add a cleanup timer that runs every minute to cleanup if no events are emitted